075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+
+The user has the flexibility to specify the [goals and the other rebalancing related configurations](https://strimzi.io/docs/operators/latest/configuring.html#type-KafkaRebalance-reference) the same way they already do today on a `KafkaRebalance` custom resource.
+This allows the user to not rely on the default Cruise Control goals configuration which would be common for any rebalance without overriding goals.
+They could define different goals based on the status of the Kafka cluster or skip hard goals in some scenarios for example.
+As already mentioned, specifying goals but more in general customizing the optimization proposal request is already possible in the current usage of a `KafkaRebalance` custom resource so it should be allowed for the auto-rebalancing as well.
+
+The idea of using a `KafkaRebalanceSettings` custom resource could be also extended to the manual rebalancing operations started by the user when they create a new `KafkaRebalance` custom resource.
+It would mean a change on the `KafkaRebalance` custom resource by adding a new `spec.settings` field to reference the rebalancing settings and deprecating all the other fields used today (i.e. `spec.goals`, ...).
+If this proposal is considered valid, doing such a change on the `KafkaRebalance` custom resource would help the proposal itself because the operator should not copy over the entire configuration from the settings resource to the rebalance resource but just set the settings reference.
+Of course, such a change would need a different proposal to be approved before this one.
+
+It is also worth mentioning that the additional `KafkaRebalanceSettings` custom resource doesn't bring any additional complexity to the operator in terms of watching, reacting and processing.
+The operator doesn't have any Kubernetes watch on the creation or update of such a resource.
+It can be only referenced by a `KafkaRebalance` custom resource so its creation doesn't trigger any action by the operator.
+The same happens when the `KafkaRebalanceSettings` custom resource is updated: no action is triggered for the operator, its updated content will be used only when a rebalancing is requested via a `KafkaRebalance` custom resource referencing those settings.
+
+### Extending the `Kafka.status` field with auto-rebalancing related state
+
+The auto-rebalancing, together with the scaling action, can take several reconciliations to finish.
+For this reason, taking into account cluster operator crashes as well, there is the need to save some state about it.
+The `Kafka` custom resource is extended by adding a new `status.cruiseControl.autoRebalance` field.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # ...
+status:
+  cruiseControl:
```

Comment:
Why do we need to nest it under the `cruiseControl` section? It is the anyway the Kafka cluster that is rebalancing. I get why the configs in `.spec` are under `.spec.cruiseControl`. But it does not seem necessary for the status.

Reply:
Yeah, I don't think it's necessary.